src/main/kotlin/com/example/springrestuser/services/PublicAssetService.kt

```
@@ -25,10 +27,27 @@ class PublicAssetService(
     }
 
     fun add(publicAssetDO: PublicAssetDO): PublicAssetDO {
-        return publicAssetStorage.add(publicAssetDO)
+        if (validate(publicAssetDO)) {
+            return publicAssetStorage.add(publicAssetDO)
+        } else throw RuntimeException("Invalid input data")
     }
 
     fun modify(publicAssetDO: PublicAssetDO) {
-        publicAssetStorage.modify(publicAssetDO)
+        if (validate(publicAssetDO)) {
+            publicAssetStorage.modify(publicAssetDO)
+        } else throw RuntimeException("Invalid input data")
+    }
+
+    fun validate(asset: PublicAssetDO): Boolean {
+        val map = mapOf(
+            AssetSubtype.A.toString() to Regex("(.{10})"),
+            AssetSubtype.B.toString() to Regex("([0-9]{9})"),
+            AssetSubtype.C.toString() to Regex("\\b[0-9a-f]{8}\\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\\b[0-9a-f]{12}\\b")
+        )
+        if (asset.externalId.matches(map[asset.type.toString()] ?: throw RuntimeException("Invalid input data"))) {
+            return true
+        }
+        return false
```

you could throw an exception here so that you won't need to introduce business logic in the controller